src/main/java/ui/components/pickers/RepositoryPicker.java

```diff
@@ -15,23 +15,23 @@
 public class RepositoryPicker {
 
     private final UI ui;
-    private final Stage stage;
     private final Consumer<String> onValueChangeCallback;
 
-    public RepositoryPicker(UI ui, Stage stage, Consumer<String> onValueChangeCallback) {
+    public RepositoryPicker(UI ui, Consumer<String> onValueChangeCallback) {
         this.ui = ui;
-        this.stage = stage;
         this.onValueChangeCallback = onValueChangeCallback;
         ui.registerEvent((ShowRepositoryPickerEventHandler) e -> Platform.runLater(() -> showRepositoryPicker()));
     }
 
     private void showRepositoryPicker() {
         Set<String> storedRepos = ui.logic.getStoredRepos();
+        new RepositoryPickerDialog(storedRepos, this::pickRepository, (repoId) -> ui.logic.isRepositoryValid(repoId));
+    }
+
+    private void pickRepository(String repoId) {
         String defaultRepo = ui.logic.getDefaultRepo();
-        RepositoryPickerDialog dialog = new RepositoryPickerDialog(storedRepos, stage);
-        Optional<String> chosenRepo = dialog.showAndWait();
-        if (chosenRepo.isPresent() && !defaultRepo.equals(chosenRepo.get())) {
-            onValueChangeCallback.accept(chosenRepo.get());
+        if (repoId != null && !defaultRepo.equals(repoId)) {
```

Comment:
Just realised the repo picker isn't a Dialog subclass, and so doesn't have access to `showAndWait`.

Instead of using null, could we pass an optional instead? That's a bit of an abuse of optionals (their use case is as return values, not parameters), but since we're not using `showAndWait` I guess it's fine.